src/compile.ts

```diff
@@ -675,21 +722,25 @@ export function compile(
           if (ts.isVariableDeclarationList(node.initializer)) {
             if (node.initializer.declarations.length === 1) {
               const varDecl = node.initializer.declarations[0];
-              if (ts.isIdentifier(varDecl.name)) {
-                // for (const i in list)
-                return newExpr(
-                  ts.isForOfStatement(node) ? "ForOfStmt" : "ForInStmt",
-                  [
-                    toExpr(varDecl, scope),
-                    toExpr(node.expression, scope),
-                    toExpr(node.statement, scope),
-                  ]
-                );
-              } else if (ts.isArrayBindingPattern(varDecl.name)) {
-                // for (const [a, b] in list)
-              }
+              return newExpr(
+                ts.isForOfStatement(node) ? "ForOfStmt" : "ForInStmt",
+                [
+                  toExpr(varDecl, scope),
+                  toExpr(node.expression, scope),
+                  toExpr(node.statement, scope),
+                ]
+              );
             }
           }
+          throw new SynthError(
+            ErrorCodes.Unsupported_Feature,
+            "For in/of loops with expression initializers are not currently supported."
+          );
+        } else if (ts.isForStatement(node)) {
+          throw new SynthError(
+            ErrorCodes.Unsupported_Feature,
+            "For loops are not currently supported. https://github.com/functionless/functionless/issues/303"
```

Comment:
I worry about this error message as it could be confused for all types of for loops.

Reply:
I called them "condition based for loops" and added `for(;;)`. We should just close #303 so we don't need to differentiate :-P.